File: java/src/main/java/com/danifoldi/dml/type/DmlNumber.java
package com.danifoldi.dml.type;

import com.danifoldi.dml.DmlSerializer;

import java.math.BigDecimal;

public class DmlNumber extends DmlCommentableValue {
    private BigDecimal value;
    private String cachedString;

    public DmlNumber(BigDecimal value) {
        this.value = value;
    }

    public DmlNumber(BigDecimal value, String cachedString) {
        this.value = value;
        this.cachedString = cachedString;
    }

    public BigDecimal value() {
        return this.value;
    }

    public void value(BigDecimal value) {
        this.value = value;
    }

    @Override
    public String serialize(int indent) {
        return DmlSerializer.serializeComment(comment(), indent) + DmlSerializer.indent(cachedString != null ? cachedString : value.toString(), indent);
    }

    @Override
    public String toString() {
        return "DmlNumber{" + value + "}";
    }
}

File: javascript/src/types/DmlNumber.js
import DmlCommentableValue from './DmlCommentableValue.js'
import DmlSerializer from '../DmlSerializer.js'

class DmlNumber extends DmlCommentableValue {
  #value
  #cachedString

  constructor(value, cachedString) {
    super()
    this.#value = value
    this.#cachedString = cachedString
  }

  value(value) {
    if (typeof value === 'undefined') {
      return this.#value
    }

    this.#value = value
    this.#cachedString = ''
  }

  serialize(indent) {
    return DmlSerializer.serializeComment(this.comment(), indent)
      + DmlSerializer.indent(
        typeof this.#cachedString !== 'undefined' && this.#cachedString !== ''
          ? this.#cachedString
          : this.#value.toString(),
        indent
      )
  }

  get [Symbol.toStringTag]() {
    return `DmlNumber{${this.#cachedString ? this.#cachedString : this.#value}}`
  }
}

export default DmlNumber
